File: src_py/aes_ops/aes_cbc.py
import os

from src_py.aes import AES


class AES_CBC:
    def __init__(self, aes: AES):
        self.aes_cbc = aes
        self.block_size = 16  # AES block size is always 16 bytes

    def encrypt(self, plaintext: bytes, key: bytes, iv: bytes = None) -> tuple:
        """
        Returns:
            tuple: (ciphertext, iv)
        """
        if iv is None:
            iv = os.urandom(self.block_size)

        if len(iv) != self.block_size:
            raise ValueError(f"IV must be {self.block_size} bytes")

        if len(key) != self.aes_cbc.key_size:
            raise ValueError(f"Key must be {self.aes_cbc.key_size} bytes")

        # Pad plaintext to block size
        padded_plaintext = pkcs7_pad(plaintext, self.block_size)

        ciphertext = b''
        previous_block = iv

        # Process each block
        for i in range(0, len(padded_plaintext), self.block_size):
            block = padded_plaintext[i:i + self.block_size]
            # XOR with previous ciphertext block (or IV for first block)
            xored_block = xor_bytes(block, previous_block)
            # Encrypt the XORed block
            encrypted_block = self.aes_cbc.encrypt(bytearray(xored_block))
            ciphertext += bytes(encrypted_block)
            previous_block = encrypted_block

        return ciphertext, iv

    def decrypt(self, ciphertext: bytes, key: bytes, iv: bytes) -> bytes:
        """
        Returns:
            bytes: Decrypted plaintext
        """
        if len(iv) != self.block_size:
            raise ValueError(f"IV must be {self.block_size} bytes")

        if len(key) != self.aes_cbc.key_size:
            raise ValueError(f"Key must be {self.aes_cbc.key_size} bytes")

        if len(ciphertext) % self.block_size != 0:
            raise ValueError("Ciphertext length must be multiple of block size")

        plaintext = b''
        previous_block = iv

        # Process each block
        for i in range(0, len(ciphertext), self.block_size):
            block = ciphertext[i:i + self.block_size]
            # Decrypt the block
            decrypted_block = self.aes_cbc.decrypt(bytearray(block), bytearray(key))
            # XOR with previous ciphertext block (or IV for first block)
            plaintext_block = xor_bytes(bytes(decrypted_block), previous_block)
            plaintext += plaintext_block
            previous_block = block

        # Remove padding
        return pkcs7_unpad(plaintext)


def encrypt_cbc(plaintext: bytes, key: bytes, iv: bytes = None) -> tuple:
    aes_instance = AES(key)
    aes_cbc = AES_CBC(aes_instance)
    return aes_cbc.encrypt(plaintext, key, iv)


def decrypt_cbc(ciphertext: bytes, key: bytes, iv: bytes) -> bytes:
    aes_instance = AES(key)
    aes_cbc = AES_CBC(aes_instance)
    return aes_cbc.decrypt(ciphertext, key, iv)


def pkcs7_pad(data: bytes, block_size: int = 16) -> bytes:
    """Apply PKCS#7 padding to data."""
    padding_length = block_size - (len(data) % block_size)
    padding = bytes([padding_length] * padding_length)
    return data + padding


def pkcs7_unpad(data: bytes) -> bytes:
    """Remove PKCS#7 padding from data."""
    if not data:
        raise ValueError("Cannot unpad empty data")
    padding_length = data[-1]
    if padding_length > len(data) or padding_length == 0:
        raise ValueError("Invalid padding")
    # Verify padding
    if data[-padding_length:] != bytes([padding_length] * padding_length):
        raise ValueError("Invalid padding")
    return data[:-padding_length]


def xor_bytes(a: bytes, b: bytes) -> bytes:
    return bytes(x ^ y for x, y in zip(a, b))

File: src_mat/aes_ops/AES_CBC.m
classdef AES_CBC

    properties
        aes         % (class AES.m)
        block_size = 16;
    end

    methods
        function obj = AES_CBC(aesObj)
            % aesObj: instance AES đã khởi tạo với key
            obj.aes = aesObj;
        end

        function [ciphertext, iv] = encrypt(obj, plaintext, iv)
            % [ciphertext, iv] = encrypt(obj, plaintext, iv)
            %
            % plaintext: uint8 hoặc char (sẽ convert sang uint8)
            % iv (optional): uint8(1x16). Nếu bỏ trống => random iv.
            %
            % ciphertext, iv: uint8 row vectors.

            pt = uint8(plaintext(:).');   % row vector

            if nargin < 3 || isempty(iv)
                % giống os.urandom(16)
                iv = uint8(randi([0 255], 1, obj.block_size));
            else
                iv = uint8(iv(:).');
                if numel(iv) ~= obj.block_size
                    error('IV must be %d bytes', obj.block_size);
                end
            end

            % PKCS#7 padding
            padded = AES_CBC.pkcs7_pad(pt, obj.block_size);

            ciphertext = zeros(1, 0, 'uint8');
            previous_block = iv;

            % xử lý từng block 16 byte
            for i = 1:obj.block_size:numel(padded)
                block = padded(i:i + obj.block_size - 1);
                % XOR với block trước đó (hoặc IV)
                xored_block = AES_CBC.xor_bytes(block, previous_block);
                % mã hóa 1 block AES
                encrypted_block = obj.aes.encrypt(xored_block);
                ciphertext = [ciphertext encrypted_block]; %#ok<AGROW>
                previous_block = encrypted_block;
            end
        end

        function plaintext = decrypt(obj, ciphertext, iv)
            % plaintext = decrypt(obj, ciphertext, iv)
            %
            % ciphertext: uint8 row vector, length multiple of 16
            % iv: uint8(1x16)

            ct = uint8(ciphertext(:).');
            iv = uint8(iv(:).');

            if numel(iv) ~= obj.block_size
                error('IV must be %d bytes', obj.block_size);
            end
            if mod(numel(ct), obj.block_size) ~= 0
                error('Ciphertext length must be multiple of block size');
            end

            previous_block = iv;
            pt = zeros(1, 0, 'uint8');

            for i = 1:obj.block_size:numel(ct)
                block = ct(i:i + obj.block_size - 1);
                % giải mã block AES
                decrypted_block = obj.aes.decrypt(block);
                % XOR với block trước (hoặc IV)
                plain_block = AES_CBC.xor_bytes(decrypted_block, previous_block);
                pt = [pt plain_block]; %#ok<AGROW>
                previous_block = block;
            end

            % bỏ padding PKCS#7
            plaintext = AES_CBC.pkcs7_unpad(pt);
        end
    end

    methods (Static, Access = private)
        function out = pkcs7_pad(data, block_size)
            data = uint8(data(:).');
            pad_len = block_size - mod(numel(data), block_size);
            if pad_len == 0
                pad_len = block_size;
            end
            pad = uint8(repmat(pad_len, 1, pad_len));
            out = [data pad];
        end

        function out = pkcs7_unpad(data)
            data = uint8(data(:).');
            if isempty(data)
                error('Cannot unpad empty data');
            end
            pad_len = double(data(end));
            if pad_len <= 0 || pad_len > numel(data)
                error('Invalid padding');
            end
            if any(data(end-pad_len+1:end) ~= data(end))
                error('Invalid padding');
            end
            out = data(1:end-pad_len);
        end

        function out = xor_bytes(a, b)
            a = uint8(a(:).');
            b = uint8(b(:).');
            if numel(a) ~= numel(b)
                error('xor_bytes: inputs must have the same length');
            end
            out = bitxor(a, b);
        end
    end
end
